File: ios/JMFTPSyncFileManager.m
//
//  JMFTPSyncFileManager.m
//  RNReactNativeFtpJm
//
//  Created by lzj on 2021/1/26.
//  Copyright © 2021 Facebook. All rights reserved.
//

#import "JMFTPSyncFileManager.h"
#import <JMSmartFTPUtils/JMFTPClient.h>
#import "JMRealFtpModel.h"

NSString *const kRNJMFTPSyncFileManager = @"kRNJMFTPSyncFileManager";

@interface JMFTPSyncFileManager () <JMFTPClientDelegate>

@property (nonatomic,assign) BOOL isHasListeners;

@property (nonatomic,copy) NSString *baseUrl;
@property (nonatomic,assign) NSInteger port;
@property (nonatomic,copy) NSString *account;
@property (nonatomic,copy) NSString *password;

@property (nonatomic,strong) NSMutableDictionary *realFtpDic;
@property (nonatomic,strong) JMFTPClient *ftpClient;
@property (nonatomic,strong) JMFTPUserInfo *userInfo;

@end

@implementation JMFTPSyncFileManager
RCT_EXPORT_MODULE(JMFTPSyncFileManager);

- (void)startObserving {
    self.isHasListeners = YES;
    [super startObserving];
}

- (void)stopObserving {
    self.isHasListeners = NO;
    [super stopObserving];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[kRNJMFTPSyncFileManager];    //添加监听方法名
}

- (NSDictionary *)constantsToExport {
    return @{kRNJMFTPSyncFileManager: kRNJMFTPSyncFileManager};    //导出监听方法名，方便JS调用
}

- (void)sendEventWithModel:(JMRealFtpModel *)model {
    if (self.isHasListeners && model) {
        NSMutableDictionary *dic = [NSMutableDictionary dictionary];
        [dic setValue:model.tag forKey:@"tag"];
        [dic setValue:model.requetUrl forKey:@"path"];
        [dic setValue:[NSNumber numberWithFloat:model.progress] forKey:@"progress"];
        NSString *jsonStr = [JMRealFtpModel toJsonString:dic];
        
        NSMutableDictionary *dataDic = [NSMutableDictionary dictionary];
        [dataDic setValue:jsonStr forKey:@"data"];
        NSString *dataStr = [JMRealFtpModel toJsonString:dataDic];
    
        [self sendEventWithName:kRNJMFTPSyncFileManager body:dataStr];
    }
}

- (BOOL)isValidInfo {
    if (self.baseUrl == nil || self.account == nil || self.password == nil) {
        return NO;
    }
    return YES;
}

- (NSMutableDictionary *)realFtpDic {
    if (!_realFtpDic) {
        _realFtpDic = [NSMutableDictionary dictionary];
    }
    return _realFtpDic;
}

#pragma mark -

RCT_EXPORT_METHOD(configFtpSyncFile:(NSString *)baseUrl mode:(NSString *)mode port:(NSInteger)port account:(NSString *)account password:(NSString *)password resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (baseUrl == nil || account == nil || password == nil) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
        
    self.baseUrl = baseUrl;
    self.port = port;
    self.account = account;
    self.password = password;
    if (resolver) resolver(nil);
}

RCT_EXPORT_METHOD(connectFTP:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (![self isValidInfo]) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数",nil);
    }
    
    _userInfo = [[JMFTPUserInfo alloc] initWithDomain:self.baseUrl port:(int)self.port account:self.account password:self.password];
    _ftpClient = [[JMFTPClient alloc] initWithUserInfo:_userInfo];
    _ftpClient.delegate = self;
    if (resolver) resolver(nil);
}

RCT_EXPORT_METHOD(findFTPFlies:(NSString *)subPath resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!self.ftpClient) {
        rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
    
    NSMutableArray *list = [NSMutableArray array];
    [self.ftpClient findFliesWithDirPath:subPath completion:^(NSArray<JMFTPFileInfo *> * _Nonnull fileInfoList) {
        for (JMFTPFileInfo *info in fileInfoList) {
            [list addObject:@[@{
                                  @"fileName": info.fileName,
                                  @"fileSize": @(info.isDir ? 0 : info.fileSize),
                                  @"filePath": info.filePath,
                                  @"fileType": @(info.isDir ? 1 : 0)
            }]];
        }
        if (resolver) resolver([JMRealFtpModel toJsonString:list]);
    }];
}

RCT_EXPORT_METHOD(downFTPFile:(NSString *)url locaUrl:(NSString *)locaUrl fileName:(NSString *)fileName tag:(NSString *)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!self.userInfo) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
    
    JMFTPClient *ftpClient = [[JMFTPClient alloc] initWithUserInfo:_userInfo];
    ftpClient.delegate = self;
    ftpClient.tag = [tag integerValue];
    
    JMRealFtpModel *model = [[JMRealFtpModel alloc] init];
    model.ftpClient = ftpClient;
    model.isDown = YES;
    model.requetUrl = url;
    model.tag = tag;
    model.resolver = resolver;
    model.rejecter = rejecter;
    [self.realFtpDic setValue:model forKey:tag];
    
    [ftpClient downloadFileWithRemotePath:url localDir:locaUrl localFileName:fileName];
    NSLog(@"开始下载：%@, tag:%@, locaUrl:%@, fileName:%@", url, tag, locaUrl, fileName);
}

RCT_EXPORT_METHOD(uploadFTPFile:(NSString *)path locaUrl:(NSString *)locaUrl fileName:(NSString *)fileName overwrite:(BOOL)overwrite tag:(NSString *)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!self.userInfo) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
    
    JMFTPClient *ftpClient = [[JMFTPClient alloc] initWithUserInfo:_userInfo];
    ftpClient.delegate = self;
    ftpClient.tag = [tag integerValue];
    
    JMRealFtpModel *model = [[JMRealFtpModel alloc] init];
    model.ftpClient = ftpClient;
    model.isDown = NO;
    model.requetUrl = path;
    model.tag = tag;
    model.resolver = resolver;
    model.rejecter = rejecter;
    [self.realFtpDic setValue:model forKey:tag];
    
    [ftpClient uploadFileWithRemotePath:path loaclPath:[NSString stringWithFormat:@"%@%@", locaUrl, fileName]];
    NSLog(@"开始上传：%@, tag:%@, locaUrl:%@, fileName:%@", path, tag, locaUrl, fileName);
}

RCT_EXPORT_METHOD(ftpPause:(NSString *)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    JMRealFtpModel *model = [self.realFtpDic objectForKey:tag];
    if (model) {
        [model pause];
        if (resolver) resolver(nil);
    } else if (rejecter) {
        rejecter(@"805", @"没有当前tag", nil);
    }
}

RCT_EXPORT_METHOD(ftpResume:(NSString *)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    JMRealFtpModel *model = [self.realFtpDic objectForKey:tag];
    if (model) {
        [model resume];
        if (resolver) resolver(nil);
    } else if (rejecter) {
        rejecter(@"805", @"没有当前tag", nil);
    }
}

RCT_EXPORT_METHOD(ftpCancel:(NSString *)tag resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    JMRealFtpModel *model = [self.realFtpDic objectForKey:tag];
    if (model) {
        [model cancel];
        [self.realFtpDic objectForKey:tag];
        if (resolver) resolver(nil);
    } else if (rejecter) {
        rejecter(@"805", @"没有当前tag", nil);
    }
}

RCT_EXPORT_METHOD(deleteFTPFile:(NSString *)path resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!self.ftpClient) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
    
    [self.ftpClient deleteFileWithPath:path completion:^(BOOL isSuccess) {
        if (isSuccess) {
            if (resolver) resolver(nil);
        } else {
            if (rejecter) rejecter(@"806", @"删除失败", nil);
        }
    }];
}

RCT_EXPORT_METHOD(moveFTPFile:(NSString *)path toPath:(NSString *)toPath overwrite:(BOOL)overwrite resolver:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    if (!self.ftpClient) {
        if (rejecter) rejecter(@"801", @"未配置ftp参数", nil);
        return;
    }
    
    [self.ftpClient moveFileFromPath:path toPath:toPath completion:^(BOOL isSuccess) {
        if (isSuccess) {
            if (resolver) resolver([JMRealFtpModel toJsonString:@{@"path": toPath}]);
        } else {
            if (rejecter) rejecter(@"807", @"移动文件失败失败", nil);
        }
    }];
}

RCT_EXPORT_METHOD(closeFTP:(RCTPromiseResolveBlock)resolver rejecter:(RCTPromiseRejectBlock)rejecter) {
    _ftpClient.delegate = nil;
    _ftpClient = nil;
    _userInfo = nil;
    [self.realFtpDic removeAllObjects];
    if (resolver) resolver(nil);
}

#pragma mark - JMFTPClientDelegate

- (void)didFtpClient:(nonnull JMFTPClient *)client progress:(nonnull NSProgress *)progress {
    JMRealFtpModel *model = [self.realFtpDic objectForKey:[NSString stringWithFormat:@"%ld", (long)client.tag]];
    if (model) {
        model.progress = progress.fractionCompleted * 1.0 / progress.totalUnitCount;
        [self sendEventWithModel:model];
    }
}

- (void)didFtpClient:(nonnull JMFTPClient *)client state:(JMFTPClientState)state {
    NSString *tag = [NSString stringWithFormat:@"%ld", (long)client.tag];
    JMRealFtpModel *model = [self.realFtpDic objectForKey:tag];
    if (model) {
        if (state == JMFTPClientStateSuccess) {
            if (model.resolver)
                model.resolver([JMRealFtpModel toJsonString:@{@"tag": tag}]);
        } else {
            if (model.rejecter)
                model.rejecter(@"804", [JMRealFtpModel toJsonString:@{@"tag": tag, @"state": [NSNumber numberWithInteger:state]}], nil);
        }
        [self.realFtpDic removeObjectForKey:tag];
    }
}

@end

File: ios/JMFTPSyncFileManager.swift
//
// ProjectName:  JiMiPlatfromProject
// FileName:     JMFTPSyncFileManager.swift
// Description:  <#Description#>
//
// Created by Jason on 2019/11/11.
// COPYRIGHT. ShenZhen JiMi Technology Co., Ltd. 2018.
// ALL RIGHTS RESERVED.
//
// No part of this publication may be reproduced, stored in a retrieval system, or transmitted,
// on any form or by any means, electronic, mechanical, photocopying, recording,
// or otherwise, without the prior written permission of ShenZhen JiMi Network Technology Co., Ltd.
//

import Foundation
import React
import JMSmartFTPUtils

struct JMRealDownUpdataFtpModel {
    
    var ftpClient:JMFTPClient? = JMFTPClient()
    var resolver: RCTPromiseResolveBlock
    var rejecter: RCTPromiseRejectBlock
    var isDown : Bool
    var progress : Double = 0
    var requetUrl : String
    var tag : String
    func pause() {
        ftpClient?.pause()
    }
    func resume()  {
        ftpClient?.resume()
    }
    func cancel()  {
        ftpClient?.cancel()
    }
}

@objc(JMFTPSyncFileManager)
class JMFTPSyncFileManager: RCTEventEmitter {
    
    let kRNFTPProgressCallback = "kRNJMFTPSyncFileManager"
    var isHasListeners = false      //JS是否有监听事件
    
    var baseUrl: String = ""
    var port: Int?
    var account: String?
    var password: String?
    
    var ftpClient: JMFTPClient?
    var userInfo: JMFTPUserInfo?
    
    var realFtpDic : [String:JMRealDownUpdataFtpModel] = [String:JMRealDownUpdataFtpModel]()
    var timer : DispatchSourceTimer?

    deinit {
        debugPrint("FTP->deinit")
    }

    /// JS开启监听事件
    override func startObserving() {
        isHasListeners = true
    }

    /// JS停止监听事件
    override func stopObserving() {
        isHasListeners = false
    }
    override func supportedEvents() -> [String]! {
        return [kRNFTPProgressCallback]
    }
    
    override func constantsToExport() -> [AnyHashable : Any]! {
        return ["kRNFTPProgressCallback": kRNFTPProgressCallback]
    }
    func sendEvent(_ model: JMRealDownUpdataFtpModel) -> Void {
        if (isHasListeners) {
            let param : [String:Any] = ["path":model.requetUrl,"tag":model.tag,"progress":model.progress]
            let jsonStr = JMFTPSyncFileTools.getJSONStringFrom(param)
            let dataStr = JMFTPSyncFileTools.getJSONStringFrom(["data":jsonStr])
            self.sendEvent(withName: kRNFTPProgressCallback, body: dataStr)
        }
    }
    
    //MARK:配置ftp
    @objc(configFtpSyncFile:mode:port:account:password:resolver:rejecter:)
    func configFtpSyncFile(baseUrl:String,
                           mode:String,
                           port:Int,
                           account:String,
                           password:String,
                           resolver:@escaping RCTPromiseResolveBlock,
                           rejecter:@escaping RCTPromiseRejectBlock)
    {
        self.baseUrl =  baseUrl
        self.port = port
        self.account = account
        self.password = password
        resolver(nil)
    }
    
    //MARK:发起ftp链接
    @objc(connectFTP:rejecter:)
    func connectFTP(resolver:@escaping RCTPromiseResolveBlock, rejecter:@escaping RCTPromiseRejectBlock) {
        guard let port = self.port,
            let account = self.account,
            let password = self.password else {
            rejecter("801","未配置ftp参数",nil)
            return
        }
        
        userInfo = JMFTPUserInfo.init(domain: self.baseUrl, port: Int32(port), account: account, password: password)
        ftpClient = JMFTPClient.init(userInfo: userInfo!)
        ftpClient?.delegate = self
        resolver(nil)
    }
    
//    "/mnt/sdcard2/DVRMEDIA/CarRecorder/PHOTO/2019_02_23"
    //MARK:获取指点文件夹下文件
    @objc(findFTPFlies:resolver:rejecter:)
    func findFTPFlies(subPath:String,
                      resolver:@escaping RCTPromiseResolveBlock,
                      rejecter:@escaping RCTPromiseRejectBlock)  {
        if subPath.count == 0 {
             resolver(JMFTPSyncFileTools.getJSONStringFrom([Any]()));
             return;
         }
        guard let ftpClient = ftpClient else {
            rejecter("801","未配置ftp参数", nil)
            return
        }
        var files = [[String: Any]]()
        
        ftpClient.findFlies(withDirPath: baseUrl + subPath, completion: { (infos) in
            for info in infos {
                files.append(["fileName": info.fileName,
                              "fileSize": info.isDir ? 0 : info.fileSize,
                              "filePath": info.filePath,
                              "fileType": info.isDir ? 1 : 0])
            }
            resolver(JMFTPSyncFileTools.getJSONStringFrom(files))
        })
    }
    
    //MARK:下载ftp文件
    @objc(downFTPFile:locaUrl:fileName:tag:resolver:rejecter:)
    func downFTPFile(url:String,
                     locaUrl:String,
                     fileName:String,
                     tag:String,
                     resolver:@escaping RCTPromiseResolveBlock,
                     rejecter:@escaping RCTPromiseRejectBlock)
    {
        guard let userInfo = userInfo else {
            rejecter("801","未配置ftp参数", nil)
            return
        }
        let ftpClient = JMFTPClient.init(userInfo: userInfo)
        ftpClient.delegate = self
        ftpClient.tag = Int(tag) ?? 0
        
        realFtpDic[tag] = JMRealDownUpdataFtpModel(ftpClient: ftpClient, resolver: resolver, rejecter: rejecter, isDown: true, progress: 0,  requetUrl: url, tag: tag)
        ftpClient.downloadFile(withRemotePath: url, localDir: locaUrl, localFileName: fileName)
        NSLog("下载开始+++++++++++")
    }
    
    //MARK:暂停下载
    @objc(ftpPause:resolver:rejecter:)
    func ftpPause(tag:String,
               resolver:@escaping RCTPromiseResolveBlock,
               rejecter:@escaping RCTPromiseRejectBlock) {
         guard let model = realFtpDic[tag] else {
            rejecter("-805","没有当前tag",nil)
            return
        }
        model.pause()
    }
    
    //MARK:恢复下载
    @objc(ftpResume:resolver:rejecter:)
    func ftpResume(tag:String,
               resolver:@escaping RCTPromiseResolveBlock,
               rejecter:@escaping RCTPromiseRejectBlock) {
         guard let model = realFtpDic[tag] else {
            rejecter("805","没有当前tag",nil)
            return
        }
        model.resume()
    }
    
    //MARK:取消下载
    @objc(ftpCancel:resolver:rejecter:)
    func ftpCancel(tag:String,
               resolver:@escaping RCTPromiseResolveBlock,
               rejecter:@escaping RCTPromiseRejectBlock) {
         guard let model = realFtpDic[tag] else {
            rejecter("805","没有当前tag",nil)
            return
        }
        model.cancel()
    }
    
    //MARK:上传文件
    @objc(uploadFTPFile:locaUrl:fileName:overwrite:tag:resolver:rejecter:)
    func uploadFTPFile(path:String,
                       locaUrl:String,
                       fileName:String,
                       overwrite:Bool,
                       tag:String,
                       resolver:@escaping RCTPromiseResolveBlock,
                       rejecter:@escaping RCTPromiseRejectBlock)  {
        guard let userInfo = userInfo else {
            rejecter("801","未配置ftp参数", nil)
            return
        }
        
        let ftpClient = JMFTPClient.init(userInfo: userInfo)
        ftpClient.delegate = self
        ftpClient.tag = Int(tag) ?? 0
        
        realFtpDic[tag] = JMRealDownUpdataFtpModel(ftpClient: ftpClient, resolver: resolver, rejecter: rejecter, isDown: false, progress: 0,  requetUrl: path, tag: tag)
        ftpClient.uploadFile(withRemotePath: path, loaclPath: locaUrl + fileName)
    }
    
    //MARK:删除文件
    @objc(deleteFTPFile:resolver:rejecter:)
    func deleteFTPFile(path:String,
                       resolver:@escaping RCTPromiseResolveBlock,
                       rejecter:@escaping RCTPromiseRejectBlock)  {
        guard let ftpClient = ftpClient else {
            rejecter("801","未配置ftp参数", nil)
            return
        }
        
        ftpClient.deleteFile(withPath: path) { (success) in
            if success {
                resolver(nil)
            } else {
                rejecter("806", "删除失败", nil)
            }
        }
    }
    
    //MARK:移动文件
    @objc(moveFTPFile:toPath:overwrite:resolver:rejecter:)
    func moveFTPFile(path:String,
                     toPath:String,
                     overwrite:Bool,
                     resolver:@escaping RCTPromiseResolveBlock,
                     rejecter:@escaping RCTPromiseRejectBlock)  {
        guard let ftpClient = ftpClient else {
            rejecter("801","未配置ftp参数", nil)
            return
        }
        
        ftpClient.moveFile(fromPath: path, toPath: toPath) { (success) in
            if success {
                resolver(JMFTPSyncFileTools.getJSONStringFrom(["path":toPath]))
            } else {
                rejecter("807","移动文件失败失败", nil)
            }
        }
    }
    
    //MARK:关闭FTP
    @objc(closeFTP:rejecter:)
    func closeFTP(resolver:@escaping RCTPromiseResolveBlock,
                  rejecter:@escaping RCTPromiseRejectBlock) {
        ftpClient?.delegate = nil
        ftpClient = nil
        userInfo = nil
        realFtpDic.removeAll()
        resolver(nil)
    }

}

extension JMFTPSyncFileManager: JMFTPClientDelegate {
    func didFtpClient(_ client: JMFTPClient, progress: Progress) {
        if realFtpDic.keys.contains(String(client.tag)),
            var model = realFtpDic[String(client.tag)]{
            model.progress = progress.fractionCompleted / Double(progress.totalUnitCount)
            self.sendEvent(model)
        }
    }
    
    func didFtpClient(_ client: JMFTPClient, state: JMFTPClientState) {
        if realFtpDic.keys.contains(String(client.tag)), let model = realFtpDic[String(client.tag)] {
            if state == .success {
                NSLog("下载结束+++++++++++")
                model.resolver(JMFTPSyncFileTools.getJSONStringFrom(["tag": String(client.tag)]))
            } else {
                model.rejecter("804", JMFTPSyncFileTools.getJSONStringFrom(["tag": String(client.tag), "errCode": state]), nil)
            }
            realFtpDic.removeValue(forKey: String(client.tag));
        }
    }
}
